staging/src/k8s.io/api/storagemigration/v1alpha1/register.go

```diff
@@ -0,0 +1,57 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// GroupName is the group name use in this package
+const GroupName = "storagemigration.k8s.io"
```

Comment:
why change the group?

Reply:
> It is to avoid potential conflict if somebody wants to run `out-of-tree` SVM along with `in-tree` for some reason.

I can also see the reasoning for built-in versus CRD type since cohabitation doesn't work for CRDs.

Reply:
I can't tell whether keeping the same API group and schema so existing API clients still work, but having all existing objects disappear when upgrading from a version serving the CRD to a version serving the built-in API is a good tradeoff or not. @deads2k, would you need to have access to both the CRD and built-in-based versions to do a seamless handoff?

The CRD-based API is alpha, but there are definitely folks using it and it's "perma-alpha" so I'm slightly more aware of what I'd tell existing users to do to switch.

Reply:
@jpbetz @deads2k What should we do here?

Reply:
> I can't tell whether keeping the same API group and schema so existing API clients still work

I would like this, but without CRD cohabitation with built-ins I don't see a logical way to make it fully seamless.

Having the controller honor both old and new during an overlap (two releases) and having the schemas match for the first transition would make it easier to make the switch to built-in types and evolve the schema from there in a second, converted alpha (v1alpha2).

A transition like this would be manageable.

1. 1.29 - v1alpha1 type in a new group that exactly matches the old schema
2. 1.29 - this particular controller uses the same lease as existing project as a secondary lease so "one at a time", allowing graceful transition
3. 1.29 - controller honors both existing alpha CRD and new alpha group
4. 1.30 - controller remove its secondary lease
5. 1.31 - controller honors only the new group
6. 1.31 - v1alpha1 and v1alpha2, with conversion between them to allow new fields to develop without insta-breaking using client

That allows someone deploying the old to start using new in 1.29 and remove their old deployment in 1.29 without a conflict or gap.  Once the 1.29 alpha controller is proven "at least as good", the new API can be used on 1.30, allowing "only new API" in 1.31 to allow versioned management transitions..

Reply:
@deads2k @jpbetz Does this group name work or something like [this](https://github.com/kubernetes/kubernetes/pull/121123#discussion_r1364198502) would make more sense?